### ISSUE TITLE
Build is littered with compiler warnings

### ISSUE THREAD

**Issue by ArcEye** [machinekit/machinekit-cnc#45]
<a href="https://github.com/zultron"><img src="https://avatars1.githubusercontent.com/u/1940736?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [zultron](https://github.com/zultron)**
_Fri Apr  6 23:20:12 2018_
_Originally opened as https://github.com/machinekit/machinekit/issues/1368_

----

The build is littered with compiler warnings.  When compiling my updates, this makes it really hard to tell if I've introduced a new warning.  #327 had them all fixed a couple of years back, but that was never merged, and it's unclear how many of those patches still address outstanding warnings, or how many would even apply.

**Comment by ArcEye** [comment]
<a href="https://github.com/ArcEye"><img src="https://avatars2.githubusercontent.com/u/770392?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ArcEye](https://github.com/ArcEye)**
_Sun Apr  8 10:29:47 2018_

----

Your original fixes were against gcc 4.9.2

I am now using gcc 7.3.0-14 and it is getting super picky, g++ even more so.

When building against some machinekit headers with Qt5, I am now having to add 
`-Wno-ignored-qualifiers` to the `QMAKE_CXXFLAGS` to get around the number of functions which declare with `static char*` but actually return `char*` and similar.

There is a set of inline macros which produce dozens of these warnings / errors (depending upon how strict your build).  They defied easy amendment and I just worked around them as above.

Certainly could do with cleaning up.

**Comment by ArcEye** [comment]
<a href="https://github.com/zultron"><img src="https://avatars1.githubusercontent.com/u/1940736?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [zultron](https://github.com/zultron)**
_Mon Apr  9 17:15:49 2018_

----

Thanks for the tips.  I'm still on gcc 5 in stretch, but I'll fix the ones I can.

**Comment by ArcEye** [comment]
Made a small start on this with machinekit/machinekit-hal#141

A large number of warnings come from the same root cause.
The coding uses `char buffer[PATH_SIZE]` as the default size for a buffer.
Later on it tries to use a line such as
```
snprintf(kmodule_dir,PATH_MAX,"%s/userland/%s", buf, uts.release);
```
What is wrong with it, is that both `kmodule_dir` and `buf` were declared as [PATH_SIZE].
It is adding `/userland/` plus `strlen(uts.release)`  to the format string length and then restricting
the string length back to PATH_MAX, to prevent buffer overruns.

So it will always get this warning, because there is a potential truncation
```
rtapi/posix/rtapi_compat.c:399:3: note: ‘snprintf’ output between 11 and 4170 bytes into a destination of size 4096
```
Don't know what to do about a lot of them, because it is ingrained in the coding style and would take a lot of work to ensure that the receiving buffer was always large enough to accommodate the full potential string size.

**Comment by zultron** [comment]
I've picked off most of the easy ones over at machinekit/machinekit#1410.

**Comment by zultron** [comment]
PR #1411 silences the `const` warnings.

There's still one more `#warning fixme` that I'm going to leave up to @ArcEye to fix.  ;)

I thought this was it for this issue, but I noticed there are additional warnings on ARM builds, and forgot @ArcEye's comment above about picky gcc v. 7.  Later....

**Comment by ArcEye** [comment]
> There's still one more #warning fixme that I'm going to leave up to @ArcEye to fix. ;)

Thank you, I shall look forward to it :smile: 

I am now on gcc 8.2.0-7, so probably even more warnings in the wings :frowning_face:

**Comment by zultron** [comment]
The last of the compiler warnings in Stretch and Jessie, all architectures, are fixed in PR machinekit/machinekit#1424.  (One was replaced with issue machinekit/machinekit-hal#162.)

In order to prevent future backslide, `./scripts/build_docker -c deb` sets `-Werror` and will now fail upon encountering compiler warnings. This may be a small surprise to both contributors and Maintainers, but the error is easy to understand:

```
Compiling hal/utils/halcmd_commands.c
hal/utils/halcmd_commands.c: In function ‘print_script_sig_info’:
hal/utils/halcmd_commands.c:2291:2: error: #warning fixme [-Werror=cpp]
 #warning fixme
  ^
cc1: all warnings being treated as errors
Makefile:407: recipe for target 'objects/hal/utils/halcmd_commands.o' failed
make: *** [objects/hal/utils/halcmd_commands.o] Error 1
```

Surely new warnings will come with Buster and gcc v. 7+, but IMO those are the subject of a new issue, and this can be closed once the PR is merged.
